mixedprecision/roots/rpoly.cpp

```diff
@@ -0,0 +1,1052 @@
+// rpoly.cpp: mixed-precision experiments with Rpoly root finding method
+//
+// Copyright (C) 2017 Stillwater Supercomputing, Inc.
+// SPDX-License-Identifier: MIT
+//
+// This file is part of the universal numbers project, which is released under an MIT Open Source license.
+// Portions of this file are adapted from work by Chris Sweeney (2015) under the 3-Clause BSD License.
+// See https://github.com/sweeneychris/RpolyPlusPlus for more details
+
+#include <iostream>
+#include <iomanip>
+
+#include <universal/number/posit/posit.hpp>
+#include <universal/number/cfloat/cfloat.hpp>
+
+#include <universal/blas/blas.hpp>
+
+template <typename Real>
+sw::universal::blas::vector<Real> Head(const sw::universal::blas::vector<Real>& vector, size_t size){
+    sw::universal::blas::vector<Real> head(size);
+    for (size_t i = 0; i < size && i < vector.size(); ++i){
+        head[i] = vector[i];
+    }
+
+    return head;
+}
+
+template <typename Real>
+sw::universal::blas::vector<Real> Tail(const sw::universal::blas::vector<Real>& vector, size_t size){
+    sw::universal::blas::vector<Real> tail(size);
+    size_t startIdx = vector.size() - size > 0 ? vector.size() - size : 0;
+    for (size_t i = startIdx; i < vector.size(); ++i){
+        tail[i] = vector[i - startIdx];
+    }
+
+    return tail;
+}
+
+template <typename Real>
+void SetZero(sw::universal::blas::vector<Real>& vector, size_t size){
+    vector.resize(size);
+    for (size_t i = 0; i < size; ++i){
+        vector[i] = 0;
+    }
+}
+
+#ifndef M_PI
+#define M_PI 3.14159265358979323846264338327950288
+#endif
+
+// Machine precision constants.
+static const double mult_eps = std::numeric_limits<double>::epsilon();
+static const double sum_eps = std::numeric_limits<double>::epsilon();
+static const double kAbsoluteTolerance = 1e-14;
+static const double kRelativeTolerance = 1e-10;
+
+enum ConvergenceType{
+    NO_CONVERGENCE = 0,
+    LINEAR_CONVERGENCE = 1,
+    QUADRATIC_CONVERGENCE = 2
+};
+
+// Evaluate the polynomial at x using the Horner scheme.
+template <typename Real>
+inline Real EvaluatePolynomial(const sw::universal::blas::vector<Real>& polynomial, const Real& x) {
+    Real v = 0.0;
+    for (size_t i = 0; i < polynomial.size(); ++i) {
+        v = v * x + polynomial[i]; // TODO: quire?
```

Comment:
v * x + p[i]  has two rounding events: the multiply and the add. You can reduce that to one with the quire, but when you do the assignment, you would lose the small values again.

the quire only works for sum of products, but here we have really have a sequence of products and the quire can't resolve that.

Order of the computation is the only dial you have to reduce catastrophic cancellation.